chain/lf3/f3.go

```diff
@@ -56,32 +60,119 @@ type F3 struct {
 type F3Params struct {
 	fx.In
 
-	PubSub       *pubsub.PubSub
-	Host         host.Host
-	ChainStore   *store.ChainStore
-	Syncer       *chain.Syncer
-	StateManager *stmgr.StateManager
-	Datastore    dtypes.MetadataDS
-	Wallet       api.Wallet
-	Config       *Config
-	LockedRepo   repo.LockedRepo
+	PubSub        *pubsub.PubSub
+	Host          host.Host
+	ChainStore    *store.ChainStore
+	Syncer        *chain.Syncer
+	StateManager  *stmgr.StateManager
+	MetaDatastore dtypes.MetadataDS
+	F3Datastore   dtypes.F3DS
+	Wallet        api.Wallet
+	Config        *Config
+	LockedRepo    repo.LockedRepo
 
 	Net api.Net
 }
 
 var log = logging.Logger("f3")
 
+var migrationKey = datastore.NewKey("/migration")
+
+func checkMigrationComplete(ctx context.Context, source datastore.Batching, target datastore.Batching) (bool, error) {
+	valSource, err := source.Get(ctx, migrationKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		log.Debug("migration not complete, no migration flag in source datastore")
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	valTarget, err := target.Get(ctx, migrationKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		log.Debug("migration not complete, no migration flag in target datastore")
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	log.Debugw("migration flags", "source", string(valSource), "target", string(valTarget))
+
+	// if the values are equal, the migration is complete
+	return bytes.Equal(valSource, valTarget), nil
+}
+
+// migrateDatastore can be removed once at least one network upgade passes
+func migrateDatastore(ctx context.Context, source datastore.Batching, target datastore.Batching) error {
+	if complete, err := checkMigrationComplete(ctx, source, target); err != nil {
+		return xerrors.Errorf("checking if migration complete: %w", err)
+	} else if complete {
+		return nil
+	}
+
+	startDate := time.Now()
+	migrationVal := startDate.Format(time.RFC3339Nano)
+
+	if err := target.Put(ctx, migrationKey, []byte(migrationVal)); err != nil {
+		return xerrors.Errorf("putting migration flag in target datastore: %w", err)
+	}
+	// make sure the migration flag is not set in the source datastore
+	if err := source.Delete(ctx, migrationKey); err != nil && !errors.Is(err, datastore.ErrNotFound) {
+		return xerrors.Errorf("deleting migration flag in source datastore: %w", err)
+	}
+
+	log.Infow("starting migration of f3 datastore", "tag", migrationVal)
+	qr, err := source.Query(ctx, query.Query{})
+	if err != nil {
+		return xerrors.Errorf("starting source wildcard query: %w", err)
+	}
+
+	// batch size of 2000, at the time of writing, F3 datastore has 150,000 keys taking ~170MiB
+	// meaning that a batch of 2000keys would be ~2MiB of memory
+	batch := autobatch.NewAutoBatching(target, 2000)
```

Comment:
Questions: 
* How long would this take?
* what's the distribution of value sizes in source? relatively flat I imagine?

Reply:
It takes 1-2s
I assume quite uniform